### ISSUE TITLE
Issue warning if the certificate name is not "contained" in the specified SNI

### ISSUE THREAD

**Issue by eeertel** [matteocorti/check_ssl_cert#78]
hi Matteo,

would it be possible to return a warning if the ${CHECKEDNAMES} is not contained in the specified SNI name ?

Background:
When specifying a SNI name we would expect a certificate name that is contained in the SNI name (wildcard handling should be considered). Otherwise a warning should be issued.

Example:
Correct:
/usr/local/nagios-commands/check_ssl_certificate.sh **--sni teamplay.siemens.com** -H webclient.eu.api.teamplay.siemens.com -r /etc/ssl/certs
SSL_CERT OK - X.509 certificate '**teamplay.siemens.com'** from 'Siemens Issuing CA Internet Server 2017' valid until Mar 21 13:30:13 2019 GMT (expires in 346 days)|days=346;;;;

Proposed warning situation:
/var/tmp/check_ssl_certificate.sh --sni **teamplay-cn.siemens.com** -H webclient.eu.api.teamplay.siemens.com -r /etc/ssl/certs
SSL_CERT WARN webclient.eu.api.teamplay.siemens.com: SNI name {teamplay-cn.siemens.com} does not match certificate name: ***.azurewebsites.net**|days=616;;;;

A possible solution would be this:

    if [ -n "${SNI}" ]; then
      NWCN=${CN#\*.}
      if echo "${SNI}" | grep -qv "${NWCN}" ; then
        warning "SNI name {$SNI} does not match certificate name: ${CN}"
      fi
    fi

Do you think this is worth adding to the check ?

tnx & bye //emil

**Comment by matteocorti** [comment]
With my version of OpenSSL I do not get anything related to `azurewebsites.net`. The server recognises the provided name and replies with the correct certificate. Which version of OpenSSL are you using? Could you please post the output with the debug (`-d`) option?

**Comment by eeertel** [comment]
here is the output of the debug output (script is modified with the snippet above)
nagios@monicinga1:/var/log/syngo/teamplay$ /usr/local/nagios-commands/check_ssl_certificate.sh --ignore-ocsp --sni teamplay.siemens.com -H webclient.us.api.teamplay.siemens.com -c 5 -r /etc/ssl/certs -w 14 -d
[DBG] ROOT_CA = -CApath /etc/ssl/certs
expect available (/usr/bin/expect)
timeout available (/usr/bin/timeout)
[DBG] perl available: /usr/bin/perl
[DBG] date available: /bin/date
found GNU date with timestamp support: enabling date computations
[DBG] check_ssl_version: 1.45.0
[DBG] OpenSSL binary: /usr/bin/openssl
[DBG] OpenSSL version: OpenSSL 1.0.1f 6 Jan 2014 (Library: OpenSSL 1.0.2g  1 Mar 2016)
[DBG] System info: Linux monicinga1 4.4.0-116-generic #140~14.04.1-Ubuntu SMP Fri Feb 16 09:25:20 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
[DBG] Date computation: GNU
[DBG] '/usr/bin/openssl s_client' supports '-servername': using -servername teamplay.siemens.com
downloading certificate to /tmp
[DBG] executing with timeout (15s): echo 'Q' | /usr/bin/openssl s_client   -connect webclient.us.api.teamplay.siemens.com:443 -servername teamplay.siemens.com -verify 6 -CApath /etc/ssl/certs    2> /tmp/check_ssl_certificate.sh9EgrwG 1> /tmp/check_ssl_certificate.shgUm4IL
[DBG]   /usr/bin/timeout 15 /bin/sh -c "echo 'Q' | /usr/bin/openssl s_client   -connect webclient.us.api.teamplay.siemens.com:443 -servername teamplay.siemens.com -verify 6 -CApath /etc/ssl/certs    2> /tmp/check_ssl_certificate.sh9EgrwG 1> /tmp/check_ssl_certificate.shgUm4IL"
[DBG] storing a copy of the retrieved certificate in webclient.us.api.teamplay.siemens.com.crt
[DBG] storing a copy of the OpenSSL errors in webclient.us.api.teamplay.siemens.com.error
parsing the certificate file
[DBG]  subject= C = DE, ST = Bayern, L = Muenchen, O = Siemens, OU = HC SV DS TP, CN = teamplay.siemens.com
[DBG] CN         = teamplay.siemens.com
[DBG] CA_O       = Siemens/serialNumber=ZZZZZZB9
[DBG] CA_CN      = Siemens Issuing CA Internet Server 2017
[DBG] SERIAL     = 7C697E7CF4003A65C58BDCE711680875B8EC740C
[DBG] OCSP_URI   = http://ocsp1.pki-services.siemens.com
[DBG] ISSUER_URI = http://ah.siemens.com/pki?ZZZZZZB9.crt
[DBG]     Signature Algorithm: sha256WithRSAEncryption
[DBG] Date computations: GNU
The certificate will expire in 337 day(s)
[DBG] Checking expiration date
[DBG] executing: /usr/bin/openssl x509 -in /tmp/check_ssl_certificate.shgUm4IL -noout -checkend 432000
[DBG] executing: /usr/bin/openssl x509 -in /tmp/check_ssl_certificate.shgUm4IL -noout -checkend 1209600
SSL_CERT OK - X.509 certificate 'teamplay.siemens.com' from 'Siemens Issuing CA Internet Server 2017' valid until Mar 21 13:30:13 2019 GMT (expires in 337 days)|days=337;14;5;;

**Comment by matteocorti** [comment]
Your version of OpenSSL was not supporting SNI correctly and this is the reason for the problem. Now we should think if it would make sense to issue a critical warning as the connection is not working as expected. I would say yes ...

**Comment by matteocorti** [comment]
Anyway in your output I don't see `azurewebsites.net`

**Comment by eeertel** [comment]
we have two (custom) bindings for this endpoint (besides the azurewebsistes default binding). 
this is the output of the second custom binding:
nagios@monicinga1:/var/log/syngo/teamplay$ /usr/local/nagios-commands/check_ssl_certificate.sh --ignore-ocsp -H webclient.us.api.teamplay.siemens.com -c 5 -r /etc/ssl/certs -w 14
SSL_CERT OK - X.509 certificate '*.us.api.teamplay.siemens.com' from 'Siemens Issuing CA Internet Server 2017' valid until Mar 16 10:30:27 2019 GMT (expires in 332 days)|days=332;14;5;;

basically we wanted to check that both bindings have been setup correctly. 
I also think that regardless of the used SSL version (we are on ubuntu 14.04 for the next 6 months or so) would help.

Thank you for your quick response !

bye //emil.

**Comment by matteocorti** [comment]
Dear Emil

I also think it makes sense, I'll install Ubuntu and test but it definitely makes sense.

Matteo

**Comment by alainwolf** [comment]
This change broke most of my http-vhost checks. 
I was under the impression, that the cn is irrelevant. 
My HTTP vhosts use DNS hostnames in subjectAltNames, 

I was unable to find a combination of command-line parameters which gave me an OK on perfectly valid certs.

The cert:
```
Subject: CN=alainwolf.ch
  X509v3 Subject Alternative Name: 
    DNS:alainwolf.ch, DNS:alainwolf.net, DNS:www.alainwolf.ch, DNS:www.alainwolf.net
```
This works:
```
$ check_ssl_cert -H 172.20.10.30 --sni alainwolf.ch
SSL_CERT OK 

```
This doesn't:
```
$ check_ssl_cert -H 172.20.10.30 --sni alainwolf.net
SSL_CERT WARN 172.20.10.30: SNI name {alainwolf.net} does not match certificate name

```
IMHO this is contradiction of RFC5280, RFC6125 and CA/Browser Forum baseline requirements. Please correct me if I am wrong.

**Comment by matteocorti** [comment]
I will roll back as soon as possible, for the moment stick to the older version. Sorry

**Comment by matteocorti** [comment]
Fixed in the e138b6e
